resources/2024-10-dapr-bindings.md

```diff
@@ -0,0 +1,368 @@
+  name: 'demoApp'
+  properties: {
+    connections: {
+      mail: {
+        source: mailing.id
+      }
+    }
+  ...
+  }
+}
+```
+
+The application can then send messages to the external system using the created binding.
+
+```go
+ctx := context.Background()
+client, _:= dapr.NewClient()
+// Sending an email with the output binding
+in := &dapr.InvokeBindingRequest{
+    Name:      CONNECTION_MAIL_COMPONENTNAME,
+    Operation: "create",
+    Data: []byte("This is the body of the message"),
+    Metadata: map[string]string{"emailTo": "example@example.net", "subject": "Not fishy"},
+}
+out, _:= client.InvokeBinding(ctx, in)
+```
+In the future, the SMTP server could be moved to a different provider or switched to a new messaging service:
+
+- For the operator, this change will only require updating a single Radius resource or recipe.
+- For the developer, the application will continue sending messages through the same binding without requiring code changes.
+
+## Design
+
+### High Level Design
+
+#### Architectural components 
+
+##### Input Binding
+```mermaid
+graph LR
+    Client -->|Request| Engine
+    Engine -->|Deploy| Kubernetes
+    Dapr_Sidecar -->|1 - Resolves| Dapr_Component_CRD
+    Dapr_Sidecar --> |2 - Watch for changes | Service_X
+    Dapr_Sidecar -->|"3 - Callback (POST)"| Application
+
+    subgraph Foreign
+        Service_X
+    end
+
+    subgraph Pod
+        Application
+        Dapr_Sidecar
+    end
+
+    subgraph Kubernetes
+        Dapr_Component_CRD
+        Pod
+    end
+```	
+
+##### Output Binding
+```mermaid
+graph LR
+    Client -->|Request| Engine
+    Engine -->|Deploy| Kubernetes
+    Application -->|Invokes| Dapr_Sidecar
+    Dapr_Sidecar -->|1 - Resolves| Dapr_Component_CRD
+    Dapr_Sidecar --> |2 - call| Service_X
+
+    subgraph Foreign
+        Service_X
+    end
+
+    subgraph Pod
+        Application
+        Dapr_Sidecar
+    end
+
+    subgraph Kubernetes
+        Pod
+        Dapr_Component_CRD
+    end
+```	
+
+#### Sequence Diagram
+
+( No changes in existing interactions)
+
+#### Common Usage Recipes
+
+Since a Binding is intended to interact with an external system outside of the operator control, a binding recipe would create only the CRD. This approach would require a two-step process to create a single CRD, which could otherwise be achieved in a single step using manual provisioning.
+
+To address this, rather than mapping [every possible binding](https://docs.dapr.io/reference/components-reference/supported-bindings/) to a recipe, the proposal is to create a few common usage recipes that cover the most frequently used cases.
+
+To maximize convenience for the user, these recipes will be designed with the following considerations:
+
+- **Local Development**: The recipe will create both the binding CRD and the "external resource".
+- **Production**: The recipe will create only the binding CRD, with the user providing the necessary information for connecting to the external resource.
```

Comment:
This is a valid usecase (pre-existing resource).

We'd usually think about the platform or infrastructure team as being the one to choose the recipes for production usage.